frontends/election-manager/src/lib/backends/backends.test.ts

```diff
@@ -253,6 +285,55 @@ function makeAdminBackend(): ElectionManagerStoreAdminBackend {
       }
 
       return { status: 404 };
+    })
+    .get('glob:/admin/elections/*/printed-ballots\\?*', async (url) => {
+      const match = url.match(
+        /^\/admin\/elections\/(.+)\/printed-ballots(\?.*)?$/
+      );
+      const electionId = match?.[1] as Id;
+      const dbEntry = db.get(electionId);
+      const query = new URLSearchParams(match?.[2] ?? '');
+
+      if (!dbEntry) {
+        return { status: 404 };
+      }
+
+      const parseQueryResult = safeParse(
+        Admin.GetPrintedBallotsQueryParamsSchema,
+        { ballotMode: query.get('ballotMode') ?? undefined }
+      );
+
+      if (parseQueryResult.isErr()) {
+        return { status: 400 };
+      }
+
+      return {
+        body: typedAs<Admin.GetPrintedBallotsResponse>({
+          status: 'ok',
+          printedBallots: await dbEntry.memoryBackend.loadPrintedBallots(
+            parseQueryResult.ok()
+          ),
+        }),
+      };
+    })
+    .post('glob:/admin/elections/*/printed-ballots', async (url, request) => {
```

Comment:
general question on this pattern: as an uninitiated engineer, it seems like unnecessary overhead to model server-side logic (and db data/side effects) in these tests vs the minimum of verifying that the client makes the expected HTTP requests. what's the team philosophy/context behind doing it this way?

Reply:
I'm not sure we have a strong/consistent philosophy on this. In general my preference is to use simplified "real" versions of any complicated component such as a "backend" (see also `MemoryStorage` and `MemoryCard`). My rationale for doing so is that it is often convenient to have such versions of the component and it often promotes better testing methodology. The alternatives also often involve opaque fixtures whose values are hardcoded mysteries.

However, I recognize that _this_ case may be taking that too far. It was easy at first, but it's become more difficult to maintain. Ditching the "storage backend" in #2717 improved things in that regard. However, I don't think we actually plan to have a demo version of this running in a browser somewhere for the general public to try, so it may be "unnecessary overhead" as you put it. I'd welcome trying out something with less overhead as long as it also avoids the "hardcoded mysteries" problem.

Reply:
ah, the `MemoryStorage`/`MemoryCard` examples are helpful! i was going to say that larger fakes like that have been useful in my experience for more behavioural testing that involves complex interactions with another component or user input, vs simple unit tests, so that's in line with what i'm seeing with those.

agreed that this particular case (at least post-refactor) could be simplified, since we can test each unit independently, but i'll see if that thought holds after i've spent a bit more time in that code!